kbase-extension/static/kbase/js/common/simpleLogViewer.js

```diff
@@ -0,0 +1,818 @@
+             */
+            renderLog() {
+                const lines = this.model.getItem('lines');
+                if (!lines || lines.length === 0) {
+                    this.ui.setContent(LOG_PANEL, 'No log entries to show.');
+                    this.renderButtons();
+                    return;
+                }
+
+                const panel = this.getLogPanel();
+                const renderedLines = panel.querySelectorAll('ol li');
+                if (renderedLines && renderedLines.length) {
+                    const olElement = panel.querySelector('ol');
+                    // append the new lines to the existing log lines
+                    const newLines = lines
+                        .slice(renderedLines.length)
+                        .map((line) => this.renderLogLine(line))
+                        .join('\n');
+                    olElement.innerHTML += newLines;
+                } else {
+                    panel.innerHTML = t('ol')(
+                        {
+                            class: `${cssBaseClass}__log_line_container`,
+                            dataElement: LOG_LINES,
+                        },
+                        lines.map((line) => this.renderLogLine(line)).join('\n')
+                    );
+                }
+
+                // if we're autoscrolling, scroll to the bottom
+                if (this.state.scrollToEndOnNext) {
+                    this.doScrollToBottom();
+                }
+                this.renderButtons();
+            }
+
+            /**
+             * Update the log control buttons
+             * Also renders the widget state if debugging mode is on
+             */
+            renderButtons() {
+                const previousStatus = (this.lastJobState && this.lastJobState.status) || null;
+                let mode = this._statusToMode(previousStatus) || 'default';
+                if (mode === 'running' && this.state.scrollToEndOnNext) {
+                    mode = 'running_scrolling';
+                }
+                this._renderButtonState(mode);
+            }
+
+            _renderButtonState(mode) {
+                // Button state
+                const { buttons } = this.buttonsByJobState[mode];
+                // the log controls section may have been removed
+                // so use try/catch to suppress any errors
+                buttons.enabled.forEach((_button) => {
+                    try {
+                        this.ui.enableButton(_button);
+                    } catch (err) {
+                        // eslint-disable-next-line no-empty
+                    }
+                });
+                buttons.disabled.forEach((_button) => {
+                    try {
+                        this.ui.disableButton(_button);
+                    } catch (err) {
+                        // eslint-disable-next-line no-empty
+                    }
+                });
+            }
+        };
+
+    const ControllerMixin = (Base) =>
```

Comment:
these mixin names are slightly arbitrary - the main purpose is to keep the different types of function separate